src/actions/requests.rs

```diff
@@ -852,27 +854,78 @@ fn location_from_racer_match(a_match: racer::Match) -> Option<Location> {
 }
 
 lazy_static! {
+    static ref NUM_THREADS: usize = ::num_cpus::get();
+
+    /// Duration of work after which we should warn something is taking a long time
+    static ref WARN_TASK_DURATION: Duration = *DEFAULT_REQUEST_TIMEOUT * 5;
+
+    /// Current work descriptions active on the work pool
+    static ref WORK: Mutex<Vec<&'static str>> = Mutex::new(vec![]);
+
     /// Thread pool for request execution allowing concurrent request processing.
     static ref WORK_POOL: rayon::ThreadPool = rayon::ThreadPool::new(
         rayon::Configuration::default()
             .thread_name(|num| format!("request-worker-{}", num))
-            // panic details will be on stderr, otherwise ignore the work panic as it
-            // will already cause a mpsc disconnect-error & there isn't anything else to log
-            .panic_handler(|_| {})
+            .num_threads(*NUM_THREADS)
     ).unwrap();
 }
 
 /// Runs work in a new thread on the `WORK_POOL` returning a result `Receiver`
 /// Panicking work will receive `Err(RecvError)` / `Err(RecvTimeoutError::Disconnected)`
-pub fn receive_from_thread<T, F>(work_fn: F) -> mpsc::Receiver<T>
+pub fn receive_from_thread<T, F>(work_fn: F, description: &'static str) -> mpsc::Receiver<T>
```

Comment:
Could you make `Description` a newtype around `&'static str`, rather than just using a string? Or use an enum (maybe better).

Reply:
I've wrapped one around `&str` for better clarity. An enum is more of a pain as the request types are not already related + I want to be able to define variants ie `WorkDescription("textDocument/definition-racer")` when we have an offshot thread from a message handler. 

I like using the language server protocol names as they are more easily identifiable outside of Rls devs (in warning logs), but also point to the problem areas just as well.